contributors/design-proposals/federated-statefulsets.md

```diff
@@ -0,0 +1,249 @@
+     containers:
+     - name: nginx
+       image: gcr.io/google\_containers/nginx-slim:0.8
+       ports:
+       - containerPort: 80
+         name: web
+       volumeMounts:
+       - name: www
+         mountPath: /usr/share/nginx/html
+ volumeClaimTemplates:
+ - metadata:
+     name: www
+   spec:
+     accessModes: ["ReadWriteOnce"]
+     resources:
+       requests:
+         storage: 1Gi
+```
+The service will be created in each of the 3 clusters unmodified.
+
+The default replica partitioning will mean a distribution of **2, 2 and 1** replicas to the 3 clusters.
+
+The statefulset spec submitted to the clusters will have the statefulset metadata.name modified to _**web+&lt;cluster-name&gt;**_.
+
+Using the example above, with statefulset name _web_, service name _nginx_ and domain name as _federation.example_ the dns names created against each instance would be as below, discoverable both locally and globally (assuming the registered cluster names being c1, c2 and c3):
+```
+web-c1-0.nginx.mynamespace.svc.federation.example
+web-c1-1.nginx.mynamespace.svc.federation.example
+web-c2-0.nginx.mynamespace.svc.federation.example
+web-c2-1.nginx.mynamespace.svc.federation.example
+web-c3-0.nginx.mynamespace.svc.federation.example
+```
+This above is fine for the case of multiple local quora (one in each cluster), but these identities cannot work across the clusters, as the local pod IPs cannot be reached across the clusters.
+
+Similar to _design alternative 1_, the proposal to make it work across the clusters is to create an additional service of type &#39;LoadBalancer&#39; for each pod instance that is created in the clusters. This will be automatically created by the federated statefulset controller, on demand, proposed to be controlled by an annotation on the statefulset.
+If the annotation specifies so (users choice), the federation controller will create the additional service of type &#39;LoadBalancer&#39; against each pod replica.
+The service name for service, one each against the pod will be the same name as the pod identity.
+The LB service creation at federation has the same drawback as discussed in _design alternative 1_.
+The proposed solution is elaborated in section **Handling additional LB Services**.
+
+## Storage volumes
+
+There are two ways a federated statefulset can be assigned a persistent storage.
+
+1 - A node local volume attached to the container
+
+2 - A persistent volume claim from the cloud provider.
+
+Either of the 2 options above can be used/specified by the user to attach a persistent store to the pods.
+
+It is known, that as of now, the same persistent volume, even if can be used in a different cluster, k8s directly does not provide an API, yet, which can aid the same. In the absence of no direct way of quick migration of the storage data from one zone to another, or from one cloud provider to another if the case be, the proposal is to disallow migration of pods across clusters.
+
+## Scale up/Scale down
```

Comment:
For the database and search systems that I can think of this would be undesirable. k8s isn't workload aware, what it means to scale up, re-shard data and all the implications of those actions. Additionally, workloads across DCs are not always homogenous or even distributed, you want to be able to scale them independently.